crates/cadenza-eval/src/ir/optimize.rs

```diff
@@ -0,0 +1,764 @@
+//! Optimization passes for the IR.
+//!
+//! This module provides a framework for running optimization passes on IR modules
+//! and implements several common optimization passes:
+//! - Constant folding: Evaluate operations on constant values at compile time
+//! - Dead code elimination: Remove instructions that produce unused values
+//! - Common subexpression elimination: Detect and eliminate redundant computations
```

Comment:
I would keep the trait and pipeline in this module. But then we should move each pass into its own module just so this file doesn't get too big.